core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -683,8 +702,7 @@ private[deploy] class Master(
       if (app.coresLeft >= coresPerExecutor) {
         // Filter out workers that don't have enough resources to launch an executor
         val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
-          .filter(worker => worker.memoryFree >= app.desc.memoryPerExecutorMB &&
-            worker.coresFree >= coresPerExecutor)
+          .filter(canLaunchExecutor(_, app.desc))
           .sortBy(_.coresFree).reverse
```

Comment:
We have an issue here if no Workers have the resources required by the executor/task requirements, then it doesn't warn/error and it doesn't retry.  Basically I started a Worker without GPUs and then said I need gpus for my executor task and it end up hanging.  I suppose one could argue this is ok as someone could start another Worker that has the resources, but I think we at least need to Warn about it

Reply:
Actually, it does retry when other executors or drivers finish. But, we can warn if executor or driver requires more resources than any of workers could have. BTW, I'm thinking do we have the same issue for memory and cores ? For example, a Worker has 10 cores at most while an executor ask for 20 cores ?

Reply:
Right if something changes (other app, other workers, etc) it retries, but if I'm the only app on the cluster its not clear why the app isn't launching.  The one thing I don't want is it to be to noisy though either. I thought about that before making the comment, because like you said if its just out of resources because other apps are running we don't really want to print anything.  I think for now we should just limit it to resources and perhaps just say no Workers are configured with the resources you requested.  If we can do that without much performance impact lets do it. If not maybe we just file a separate jira for it and look at it there

Reply:
How about this way?

```
for (app <- waitingApps) {
   ...
   val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
     .filter(canLaunchExecutor(_, app.desc))
     .sortBy(_.coresFree).reverse
   if (waitingApps.size == 1 && usableWorkers.isEmpty) {
    logWarn("The app requires more resources(mem, core, accelerator) than any of Workers could have .") 
  }
  ...
}
```

Telling the Workers are not configured with the resources as an app requested may require more changes. For example, you may need to traversal workers again to judge whether it's due to resources(I mean accelerator) or memory or cores. Or, you need to refactor `canLaunchExecutor` to tell more details.

Reply:
that sounds good for now.  Lets also leave it called resource since that is what its called everywhere right now.  just leave off the (mem, core, accelerator) part.